### PR TITLE
Update webpack to v5.x

### PR DESCRIPTION
**Pull request by walbo**
Trac ticket: https://core.trac.wordpress.org/ticket/51750

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: tools/webpack/shared.js
Comment (gziolo) [inline tools/webpack/shared.js:17]: Can you explain why `moduleIds` is different in webpack 5? Did they change the name?

Comment (gziolo) [reply tools/webpack/shared.js:17]: Correct. I found an answer: https://webpack.js.org/configuration/optimization/#optimizationmoduleids.

**Comment by gziolo** [comment]
Great work. I'll have a closer look next week. Are there more differences like the one I commented about in https://github.com/WordPress/wordpress-develop/pull/2263#discussion_r799210841? I know that `CopyWebpackPlugin` has a different API with webpack 5, anything else that comes to your mind?

**Comment by walbo** [comment]
> Great work. I'll have a closer look next week. Are there more differences like the one I commented about in [#2263 (comment)](https://github.com/WordPress/wordpress-develop/pull/2263#discussion_r799210841)? I know that `CopyWebpackPlugin` has a different API with webpack 5, anything else that comes to your mind?

Other than `CopyWebpackPlugin`, the biggest difference is the removal of `@wordpress/custom-templated-path-webpack-plugin` and `@wordpress/library-export-default-webpack-plugin`. This can now be fixed directly in the [`entry`](https://github.com/WordPress/wordpress-develop/pull/2263/files#diff-47f938308062dccd6b66f867fd9030b08c2f582b8e2e28351b999d84360d871dL160-R144) config.

File: tools/webpack/shared.js
Comment (walbo) [reply tools/webpack/shared.js:17]: Yes, they have deprecated `hashed` in favor of `deterministic`. In the migration guide they mention that you should consider to remove `moduleIds` if you use that (https://webpack.js.org/migrate/5#clean-up-configuration). Need to read trough https://core.trac.wordpress.org/ticket/53192 and understand why it was added in the first place.

Comment (gziolo) [reply tools/webpack/shared.js:17]: Makes sense 👍🏻

**Comment by gziolo** [comment]
@walbo, any chances to bring this PR up to date? I'd love to land that before updating WordPress packages in core for the upcoming WordPress 6.0 release.

**Comment by walbo** [comment]
@gziolo Yes, I can do that. Will try to make it happen later today, if not I'll to it tomorrow.

**Comment by gziolo** [comment]
This PR will unblock https://github.com/WordPress/wordpress-develop/pull/2519 which brings React Fast Refresh support to WordPress core for block development with `@wordpress/scripts`.

File: tools/webpack/shared.js
Comment (walbo) [inline tools/webpack/shared.js:42]: Not sure why v5 shows alot more stats, but changed it to only show errors. From the logs I only see v4 logging an hash and the webpack version.

File: tools/webpack/shared.js
Comment (walbo) [inline tools/webpack/shared.js:20]: To keep the License comments inside the minimized files we need to set this.

Omitting this will create a `.min.LICENCE.txt` file that contains the licenses. See https://webpack.js.org/plugins/terser-webpack-plugin/#extractcomments

Would be great to enable this to make the minimized files smaller, but since they are inlined today I opted to still have them inline.

**Comment by walbo** [comment]
@gziolo The PR should be up to date. Added a couple of comments.

**Comment by gziolo** [comment]
We can use the PR with webpack 5 upgrade in Gutenberg as a reference: https://github.com/WordPress/gutenberg/pull/33818

File: tools/webpack/shared.js
Comment (gziolo) [inline tools/webpack/shared.js:52]: Nice refactoring here 👍🏻

File: tools/webpack/shared.js
Comment (gziolo) [reply tools/webpack/shared.js:20]: This is what we have in Gutenberg:

https://github.com/WordPress/gutenberg/blob/a618eaf35a417cd4cb7492b324cd6b8e5b86b7d9/tools/webpack/shared.js#L26-L42

I think the part for translation comments was necessary after migration to v5.

Comment (gziolo) [reply tools/webpack/shared.js:20]: Related PRs: https://github.com/WordPress/gutenberg/pull/22990 and https://github.com/WordPress/gutenberg/pull/28231. @swissspidy, do we need the same handling in WordPress core? We have non-minified files included, too.

**Review by gziolo — APPROVED** [review]
Code wise it looks good. It simply needs more testing to ensure we don't break any existing workflows. I would appreciate some sanity check from someone else.

File: tools/webpack/shared.js
Comment (swissspidy) [reply tools/webpack/shared.js:20]: `extractComments: false` makes sense.

Using the same config as Gutenberg to keep translator comments and avoid mangling i18n functions also makes sense

File: tools/webpack/shared.js
Comment (gziolo) [reply tools/webpack/shared.js:42]: Should we include also warnings: `'errors-warnings'`? Unless it's too noisy. We use the default value in Gutenberg.

**Comment by paaljoachim** [comment]
I noticed that this PR is included in the [WordPress 6.0 Project board](https://github.com/orgs/WordPress/projects/27/views/1). 
I assume based on @gziolo comment that it should be removed from the project board. Please remove it if that is the case.

**Comment by gziolo** [comment]
> I noticed that this PR is included in the [WordPress 6.0 Project board](https://github.com/orgs/WordPress/projects/27/views/1).
> I assume based on @gziolo comment that it should be removed from the project board. Please remove it if that is the case.

Yes, it's correctly included. I asked the Build/Test Tools component maintainers to run additional checks with the intention to land it before WordPress 6.0 Beta 1.

**Comment by paaljoachim** [comment]
Thanks for the status update, Greg!

**Comment by gziolo** [comment]
I'm preparing the final patch to land in WordPress core so I opened https://github.com/WordPress/wordpress-develop/pull/2561 to ensure that CI checks remain green after rebasing and resolving conflicts with `trunk`.

**Comment by gziolo** [comment]
Committed with https://core.trac.wordpress.org/changeset/53135. Major props to @walbo for making it happen! 🎉

**Pull request closed without merging**
